include/picongpu/fields/MaxwellSolver/YeePML/YeePML.hpp

```diff
@@ -0,0 +1,431 @@
+                    if( exchange >= BACK )
+                        axis = 2;
+                    uint32_t direction = exchange % 2;
+
+                    // No PML at the borders between two local domains
+                    bool hasNeighbour = communicationMask.isSet( exchange );
+                    if( hasNeighbour )
+                        localThickness( axis, direction ) = 0;
+
+                    // Disable PML during laser initialization
+                    if( fields::laserProfiles::Selected::initPlaneY == 0 )
+                    {
+                        bool isLaserInitializationOver =
+                            (currentStep * DELTA_T) >= fields::laserProfiles::Selected::INIT_TIME;
+                        if( numSlides == 0 && !isLaserInitializationOver && exchange == TOP )
+                            localThickness( axis, direction ) = 0;
+                    }
+
+                    // Disable PML at the far side of the moving window
+                    if( movingWindow.isSlidingWindowActive( currentStep ) && exchange == BOTTOM )
+                        localThickness( axis, direction ) = 0;
+                }
+                return localThickness;
+            }
+
+            //! Verify that PML fits the local domain
+            void checkLocalThickness( Thickness const localThickness ) const
+            {
+                auto const localDomain = Environment< simDim >::get( ).SubGrid( ).getLocalDomain( );
+                auto const localPMLSize = localThickness.negativeBorder + localThickness.positiveBorder;
+                auto pmlFitsDomain = true;
+                for( uint32_t dim = 0u; dim < simDim; dim++ )
+                    if( localPMLSize[ dim ] > localDomain.size[ dim ] )
+                        pmlFitsDomain = false;
+                if( !pmlFitsDomain )
+                    throw std::out_of_range( "Requested PML size exceeds the local domain" );
+            }
+
+            //! Get number of workers for kernels
+            static constexpr uint32_t getNumWorkers( )
+            {
+                return pmacc::traits::GetNumWorkers<
+                    pmacc::math::CT::volume< SuperCellSize >::type::value
+                >::value;
+            }
+
+        };
+
+    } // namespace detail
+    } // namespace yeePML
+
+    /** Yee field solver with perfectly matched layer (PML) absorber
+     *
+     * Absorption is done using convolutional perfectly matched layer (CPML),
+     * implemented according to [Taflove, Hagness].
+     *
+     * This class template is a public interface to be used, e.g. in .param
+     * files and is compatible with other field solvers. Parameters of PML
+     * are taken from pml.param, pml.unitless.
+     *
+     * Enabling this solver results in more memory being used on a device:
+     * 12 additional scalar field values per each grid cell of a local domain.
+     * Another limitation is not full persistency with checkpointing: the
```

Comment:
This is a quite hard limitation. @psychocoderHPC, @ax3l and @steindev what do you think about that? Fore me this limits usability for long simulations where restarts are required. However, usually those are the ones that need better absorbers.

Reply:
Yes, these limitations will be lifted in following PRs. The plan for that is given in the first message in this discussion.